docs/src/content/docs/book/security-best-practices.mdx

```diff
@@ -0,0 +1,435 @@
+
+##### Do's ✅
+
+Perform string parsing from human-readable formats into machine-readable binary structures **off-chain** to keep the contract efficient.
+
+```tact
+message Sample {
+    parsedField: Slice;
+}
+
+receive(msg: Sample) {
+    // Process msg.parsedField directly
+}
+```
+
+##### Don'ts ❌
+
+Avoid parsing strings from human-readable formats into binary structures **on-chain**, as it increases computational overhead and gas costs.
+
+```tact
+message Sample {
+    field: String;
+}
+
+receive(msg: Sample) {
+    // Parsing occurs on-chain, which is inefficient
+    let parsed = field.fromBase64();
+}
+```
+
+## Gas limitation
+
+Be careful with the `Out of gas error`. It cannot be handled, so try to precalculate the gas consumption for each receiver using tests, if possible.
+This will help not to waste extra gas, because the transaction will fail anyway.
+
+##### Do's ✅
+
+```tact
+message Vote {
+    votes: Int as int32;
+}
+
+contract Sample2 {
+    const voteGasUsage = 10000; // precompute with tests
+
+    receive(msg: Vote) {
+        require(context().value > getComputeFee(self.voteGasUsage, false), "Not enough gas!");
+    }
+}
+```
+
+## Identity validation
+
+Always validate identity of the sender if needed. This may be achieved through [`Ownable{:tact}`](/ref/stdlib-ownable) trait or using state init validation. You can read more about [Jetton validation](/cookbook/jettons#accepting-jetton-transfer) and [NFT validation](/cookbook/nfts#accepting-nft-ownership-assignment).
+
+##### Do's ✅
+
+Use [`Ownable{:tact}`](/ref/stdlib-ownable) trait.
+
+```tact
+import "@stdlib/ownable";
+
+contract Counter with Ownable {
+    owner: Address;
+    val: Int as uint32;
+
+    init() {
+        self.owner = address("OWNER_ADDRESS");
+        self.val = 0;
+    }
+
+    receive("admin-double") {
+        self.requireOwner();
+        self.val = self.val * 2;
+    }
+}
+```
+
+##### Don'ts ❌
+
+Don't execute a message without validating the sender's identity!
+
+```tact
+contract Example with Deployable {
+    myJettonWalletAddress: Address;
+    myJettonAmount: Int as coins = 0;
+
+    init(jettonWalletCode: Cell, jettonMasterAddress: Address) {
+        self.myJettonWalletAddress = calculateJettonWalletAddress(
+            myAddress(),
+            jettonMasterAddress,
+            jettonWalletCode,
+        );
+    }
+
+    receive(msg: JettonTransferNotification) {
+        self.myJettonAmount += msg.amount;
+    }
+}
+```
+
+## Replay protection
+
+Replay protection is a security mechanism to prevent an attacker from reusing a previous message. More about replay protection may be found [here](https://docs.ton.org/develop/smart-contracts/guidelines/external-messages).
+
+##### Do's ✅
+
+Always include and validate a unique identifier, like seqno, to differentiate messages. Update the identifier after successful processing to prevent duplicates.
```

Comment:
Or, you can implement replay protection like it is done in Highload V3. It is not seqno-based